overlord/snapstate/handlers.go

```diff
@@ -1404,6 +1429,9 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) (err erro
 		}
 		err = m.backend.UnlinkSnap(oldInfo, linkCtx, NewTaskProgressAdapterLocked(t))
 		if err != nil {
+			if relinkErr := m.restoreUnlinkOnError(t, oldInfo, perfTimings); relinkErr != nil {
+				return relinkErr
+			}
```

Comment:
hm ideally we should use errors.Join(), but since our Go is too old, I'd return the first error as the main reason for failure and log the relink error in the task:

```suggestion
			if relinkErr := m.restoreUnlinkOnError(t, oldInfo, perfTimings); relinkErr != nil {
				t.Errorf("cannot restore unlinked snap: %v", relinkErr)
			}
```